### PR TITLE
Fix for bug #7640  run is not working when project already has a misconfigured tsconfig.json

### PR DESCRIPTION
**Pull request by praveendvd**
## Proposed changes

Made changes to move creation of wdio.config and tsconfig to test folder if a tsconfig file already exists under the root project. Fix for [🐛 Bug]: run is not working when project already has a misconfigured tsconfig.json #7640 

## Types of changes

[//]: # (What types of changes does your code introduce to WebdriverIO?)
[//]: # (_Put an `x` in the boxes that apply_)

- [X] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

## Checklist

[//]: # (_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._)

- [X] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/main/CONTRIBUTING.md) doc
- [X] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [X] I have added proper type definitions for new commands (if appropriate)

## Further comments

Have to write tests

### CODE REVIEW

File: packages/wdio-cli/src/commands/config.ts [hunk 1]
```diff
@@ -182,7 +189,9 @@ const runConfig = async function (useYarn: boolean, yes: boolean, exit = false)
     }
 
     try {
-        await renderConfigurationFile(parsedAnswers)
+        const configPath = hasFile('test/tsconfig.json') ? 'test' : undefined
+        answers.generateTSConfigFile ? await renderConfigurationFile(parsedAnswers, configPath) :
+            await renderConfigurationFile(parsedAnswers)
```
Comment (christian-bromann) [inline packages/wdio-cli/src/commands/config.ts:194]:
```suggestion
        await renderConfigurationFile(
            parsedAnswers,
            answers.generateTSConfigFile && hasFile('test/tsconfig.json') ? 'test' : undefined
        )
```

File: packages/wdio-cli/src/constants.ts [hunk 2]
```diff
@@ -21,7 +21,7 @@ WDIO Configuration Helper
 export const CONFIG_HELPER_SUCCESS_MESSAGE = `
 Configuration file was created successfully!
 To run your tests, execute:
-$ npx wdio run wdio.conf.%s
+$ npx wdio run <path>/wdio.conf.%s
```
Comment (christian-bromann) [inline packages/wdio-cli/src/constants.ts:24]:
```suggestion
$ npx wdio run %s/wdio.conf.%s
```

Comment (christian-bromann) [reply packages/wdio-cli/src/constants.ts:24]: Then we need to update the `format` function where this const is used to add the path if conditions are met.

File: packages/wdio-cli/src/commands/config.ts [hunk 3]
```diff
@@ -77,11 +77,18 @@ const runConfig = async function (useYarn: boolean, yes: boolean, exit = false)
...
+                    path.join(process.cwd(), 'tsconfig.json'),
+                    JSON.stringify(config, null, 4)
+                )
+            } else {
+                fs.ensureDirSync(path.join(process.cwd(), 'test'))
+                await fs.promises.writeFile(
+                    path.join(process.cwd(), 'test/tsconfig.json'),
+                    JSON.stringify(config, null, 4)
+                )
+            }
```
Comment (christian-bromann) [inline packages/wdio-cli/src/commands/config.ts:91]: we can simplify this via:

```suggestion
            await fs.promises.writeFile(
                path.join(process.cwd(), (hasFile('tsconfig.json') ? 'test' : ''), 'tsconfig.json'),
                JSON.stringify(config, null, 4)
            )
```

**Comment by praveendvd** [comment]
@christian-bromann  I am not sure why the type failure here, is there anything i need to change in my commit ? it looks like non commit related issue

File: packages/wdio-cli/src/templates/wdio.conf.tpl.ejs [hunk 4]
```diff
@@ -4,6 +4,34 @@
     // Runner Configuration
     // ====================
     //<%
+    if(answers.generateTSConfigFile && hasFile('test/tsconfig.json')){ %>
```
Comment (Badisi) [inline packages/wdio-cli/src/templates/wdio.conf.tpl.ejs:7]: Are we not supposed to generate this part no matter where the `tsconfig.json` is ?

For instance, if it's a new project, the `tsconfig.json` will be at the root of the project but it would still require `autoCompileOpts`, don't you think ?

Comment (praveendvd) [reply packages/wdio-cli/src/templates/wdio.conf.tpl.ejs:7]: It picks up that automatically if it's in root folder , only if it's in non root location then only we need to specify the tsconfig location

File: packages/wdio-cli/src/commands/config.ts [hunk 5]
```diff
@@ -78,8 +78,9 @@ const runConfig = async function (useYarn: boolean, yes: boolean, exit = false)
                 }
             }
 
+            fs.ensureDirSync(path.join(process.cwd(), 'test'))
             await fs.promises.writeFile(
-                path.join(process.cwd(), 'tsconfig.json'),
+                path.join(process.cwd(), (hasFile('tsconfig.json') ? 'test' : ''), 'tsconfig.json'),
```
Comment (christian-bromann) [inline packages/wdio-cli/src/commands/config.ts:83]: I am seeing `hasFile('tsconfig.json')` here and in line 186, let's create a variable so we only need to call this function once.

File: packages/wdio-cli/src/constants.ts [hunk 6]
```diff
@@ -284,6 +284,12 @@ export const QUESTIONNAIRE = [{
         : hasFile('tsconfig.json')
             ? COMPILER_OPTIONS.ts // default to TypeScript
             : COMPILER_OPTIONS.nil // default to no compiler
+}, {
+    type: 'confirm',
+    name: 'generateTSConfigFile',
+    message: 'Do you want WebdriverIO to autogenerate tsconfig file?',
```
Comment (christian-bromann) [inline packages/wdio-cli/src/constants.ts:290]: I wonder if this is the correct question to ask here. What happened if a user says `no`? They would see a bunch of TS errors due to a missing tsconfig, right?

File: packages/wdio-cli/src/commands/config.ts [hunk 7]
```diff
@@ -65,7 +65,7 @@ const runConfig = async function (useYarn: boolean, yes: boolean, exit = false)
         if (!hasPackage('ts-node')) {
             packagesToInstall.push('ts-node', 'typescript')
         }
-        if (!hasFile('tsconfig.json')){
+        if (answers.generateTSConfigFile) {
```
Comment (christian-bromann) [inline packages/wdio-cli/src/commands/config.ts:68]: So I suggest to remove this if statement and always generate a tsconfig file.

**Review by christian-bromann — COMMENTED** [review]
To simplify the code, how about we say: always generate a tsconfig to `./test/tsconfig.json`?

File: packages/wdio-cli/src/constants.ts (hunk 6 above)
Comment (praveendvd) [reply packages/wdio-cli/src/constants.ts:290]: @christian-bromann  Yes thats why we added automatic creation of it ( in previous implementation ) . IN this case there is a console message saying what needs to be added to the tsconfig file so if user types  no still we are giving this console message so user have to add those to tsconfig file .

File: packages/wdio-cli/src/commands/config.ts (hunk 7 above)
Comment (praveendvd) [reply packages/wdio-cli/src/commands/config.ts:68]: @christian-bromann 

what if there already exists a tsconfig file as the main issue mentioned. we need that check to know whether to create the tsconfig in current directory or inside test

File: packages/wdio-cli/src/constants.ts (hunk 6 above)
Comment (christian-bromann) [reply packages/wdio-cli/src/constants.ts:290]: Why do we need this question then?

File: packages/wdio-cli/src/commands/config.ts (hunk 5 above)
Comment (praveendvd) [reply packages/wdio-cli/src/commands/config.ts:83]: @christian-bromann  will do that 🥳

File: packages/wdio-cli/src/commands/config.ts (hunk 7 above)
Comment (christian-bromann) [reply packages/wdio-cli/src/commands/config.ts:68]: Why don't we just always create a custom tsconfig file in `./test`? I think it makes the codebase easier and for the user it doesn't really matter, right?

File: packages/wdio-cli/src/constants.ts (hunk 6 above)
Comment (praveendvd) [reply packages/wdio-cli/src/constants.ts:290]: @christian-bromann  now we removed the automatic creation , if user says yes then we creates the file in current or test folder depending on availability of already existing tsconfig file. else we wont do anything . do you want to remove the question ?

File: packages/wdio-cli/src/commands/config.ts (hunk 7 above)
Comment (praveendvd) [reply packages/wdio-cli/src/commands/config.ts:68]: @christian-bromann  As mentioned before, if we create inside test , then wdio.config should be also inside test folder . So the structure ts project looks different than for all other compilers . 

in our previous comment it was decided that we will create inside test folder only if a tsconfig file exists in the root directory

Comment (praveendvd) [reply packages/wdio-cli/src/commands/config.ts:68]: https://github.com/webdriverio/webdriverio/issues/7640#issuecomment-972694037

Comment (christian-bromann) [reply packages/wdio-cli/src/commands/config.ts:68]: > then wdio.config should be also inside test folder . So the structure ts project looks different than for all other compilers 

We can just set [`autoCompileOpts `](https://webdriver.io/docs/typescript#configuration) to point to the correct `tsconfig.json` as you already implemented it in this PR.

> in our previous comment it was decided that we will create inside test folder only if a tsconfig file exists in the root directory

Correct.

Comment (praveendvd) [reply packages/wdio-cli/src/commands/config.ts:68]: @christian-bromann 

> in our previous comment it was decided that we will create inside test folder only if a tsconfig file exists in the root directory - correct

This is how its implemented now with autocompile option be added if tsconfig file is not in root . So i will keep it the way it is ? or you want tsconfig to be in test folder in all cases ?

File: packages/wdio-cli/src/constants.ts (hunk 6 above)
Comment (christian-bromann) [reply packages/wdio-cli/src/constants.ts:290]: Yes, I think we clarified in https://github.com/webdriverio/webdriverio/issues/7640#issuecomment-972694037 that we always want to create a tsconfig file. Now I also suggest to put the tsconfig file always in `./test` because it simplifies the implementation and has no difference for the user. Then we don't need this question.

File: packages/wdio-cli/src/commands/config.ts (hunk 7 above)
Comment (christian-bromann) [reply packages/wdio-cli/src/commands/config.ts:68]: > or you want tsconfig to be in test folder in all cases ?

Let's always put it in `/test` because it simplifies the implementation and it makes no difference for the user.

File: packages/wdio-cli/src/constants.ts (hunk 6 above)
Comment (praveendvd) [reply packages/wdio-cli/src/constants.ts:290]: @christian-bromann will do that 🥳

Comment (christian-bromann) [reply packages/wdio-cli/src/constants.ts:290]: Awesome, sorry if I wasn't clear enough at the beginning.

File: packages/wdio-cli/src/commands/config.ts [hunk 8]
```diff
@@ -78,8 +101,9 @@ const runConfig = async function (useYarn: boolean, yes: boolean, exit = false)
                 }
             }
 
+            fs.ensureDirSync(path.join(process.cwd(), 'test'))
```
Comment (christian-bromann) [inline packages/wdio-cli/src/commands/config.ts]: if we can use `await`, we should use it
```suggestion
            await fs.ensureDir(path.join(process.cwd(), 'test'))
```

File: packages/wdio-cli/src/commands/config.ts [hunk 9]
```diff
@@ -58,14 +58,37 @@ const runConfig = async function (useYarn: boolean, yes: boolean, exit = false)
...
+    const parsedPaths = getPathForFileGeneration(answers)
+
+    const parsedAnswers: ParsedAnswers = {
+        ...answers,
+        runner: runnerPackage.short as 'local',
+        framework: frameworkPackage.short,
+        reporters: reporterPackages.map(({ short }) => short),
+        services: servicePackages.map(({ short }) => short),
+        packagesToInstall,
+        isUsingTypeScript: answers.isUsingCompiler === COMPILER_OPTIONS.ts,
```
Comment (christian-bromann) [inline packages/wdio-cli/src/commands/config.ts:74]:
```suggestion
        isUsingTypeScript: answers.isUsingCompiler === COMPILER_OPTIONS.ts,
        tsConfigFilePath: path.join(process.cwd(), 'test', 'tsconfig.json'),
```

File: packages/wdio-cli/src/commands/config.ts [hunk 10]
```diff
@@ -78,8 +101,9 @@ const runConfig = async function (useYarn: boolean, yes: boolean, exit = false)
                 }
             }
 
+            fs.ensureDirSync(path.join(process.cwd(), 'test'))
             await fs.promises.writeFile(
-                path.join(process.cwd(), 'tsconfig.json'),
+                path.join(process.cwd(), 'test', 'tsconfig.json'),
```
Comment (christian-bromann) [inline packages/wdio-cli/src/commands/config.ts]:
```suggestion
                parsedAnswers.tsConfigFilePath,
```

File: packages/wdio-cli/src/commands/config.ts [hunk 11]
```diff
@@ -158,31 +182,11 @@ const runConfig = async function (useYarn: boolean, yes: boolean, exit = false)
...
-        destPageObjectRootPath: parsedPaths.destPageObjectRootPath,
-        relativePath : parsedPaths.relativePath
-    }
-
     try {
-        await renderConfigurationFile(parsedAnswers)
+        await renderConfigurationFile(
+            parsedAnswers,
+            parsedAnswers.isUsingTypeScript ? 'test' : undefined
+        )
```
Comment (christian-bromann) [inline packages/wdio-cli/src/commands/config.ts:189]:
```suggestion
        await renderConfigurationFile(parsedAnswers)
```

File: packages/wdio-cli/src/utils.ts [hunk 12]
```diff
@@ -218,11 +218,11 @@ export function convertPackageHashToObject(pkg: string, hash = '$--$'): Supporte
     }
 }
 
-export async function renderConfigurationFile (answers: ParsedAnswers) {
+export async function renderConfigurationFile (answers: ParsedAnswers, configPath = '') {
```
Comment (christian-bromann) [inline packages/wdio-cli/src/utils.ts:221]:
```suggestion
export async function renderConfigurationFile (answers: ParsedAnswers) {
```

File: packages/wdio-cli/src/utils.ts [hunk 13]
```diff
@@ -218,11 +218,11 @@ export function convertPackageHashToObject(pkg: string, hash = '$--$'): Supporte
...
     const tplPath = path.join(__dirname, 'templates/wdio.conf.tpl.ejs')
     const filename = `wdio.conf.${answers.isUsingTypeScript ? 'ts' : 'js'}`
     const renderedTpl = await renderFile(tplPath, { answers })
-    return fs.promises.writeFile(path.join(process.cwd(), filename), renderedTpl)
+    return fs.promises.writeFile(path.join(process.cwd(), configPath, filename), renderedTpl)
```
Comment (christian-bromann) [inline packages/wdio-cli/src/utils.ts:225]:
```suggestion
    return fs.promises.writeFile(parsedAnswers.tsConfigFilePath, renderedTpl)
```

**Review by christian-bromann — COMMENTED** [review]
More simplifcations

File: packages/wdio-cli/src/utils.ts (hunk 13 above)
Comment (praveendvd) [reply packages/wdio-cli/src/utils.ts:225]: @christian-bromann  parsedAnswers.tsConfigFilePath will have somepath/tsconfig.json , how will we pass this to write File of config file ? i think we need to keep the path.join here

Comment (christian-bromann) [reply packages/wdio-cli/src/utils.ts:225]: ah right, nevermind then

File: packages/wdio-cli/src/commands/config.ts [hunk 14]
```diff
@@ -158,31 +183,10 @@ const runConfig = async function (useYarn: boolean, yes: boolean, exit = false)
...
-        destSpecRootPath: parsedPaths.destSpecRootPath,
-        destPageObjectRootPath: parsedPaths.destPageObjectRootPath,
-        relativePath : parsedPaths.relativePath
-    }
-
     try {
-        await renderConfigurationFile(parsedAnswers)
+        await renderConfigurationFile(
+            parsedAnswers
+        )
```
Comment (christian-bromann) [inline packages/wdio-cli/src/commands/config.ts:189]:
```suggestion
        await renderConfigurationFile(parsedAnswers)
```

**Review by christian-bromann — APPROVED** [review]
👍

**Comment by christian-bromann** [comment]
@praveendvd can you check the failing unit tests?

**Comment by praveendvd** [comment]
@christian-bromann  the error is cannot read writeFile of undefined , but the fs.promise.writefile is already mocked in test (haven't changed anything )  . I am not sure why this error is thrown. 

Do you have any suggestions else I will debug tomorrow

**Comment by praveendvd** [comment]
@christian-bromann  had to change the import fs to fs-extra will fix it

File: packages/wdio-cli/tests/commands/__snapshots__/config.test.ts.snap [hunk 15]
```diff
@@ -299,6 +299,7 @@ Object {
     "services": Array [
       undefined,
     ],
+    "tsConfigFilePath": "/mnt/d/MyProjects/webdriverio/all/linux/webdriverio/test/tsconfig.json",
```
Comment (praveendvd) [inline packages/wdio-cli/tests/commands/__snapshots__/config.test.ts.snap:302]: "tsConfigFilePath": "/home/runner/work/webdriverio/webdriverio/test/tsconfig.json",

Comment (praveendvd) [inline packages/wdio-cli/tests/commands/__snapshots__/config.test.ts.snap:302]:
```suggestion
    "tsConfigFilePath": "/home/runner/work/webdriverio/webdriverio/test/tsconfig.json",
```

**Comment by praveendvd** [comment]
@christian-bromann  The snap shot of linux has 

  "tsConfigFilePath": "/mnt/d/MyProjects/webdriverio/all/linux/webdriverio/test/tsconfig.json",


and for 

windows 

    "tsConfigFilePath": "D:\\a\\webdriverio\\webdriverio\\test\\tsconfig.json",


should i create a snapshot resolver ? or any other quick fix for this thats already implemented

**Comment by christian-bromann** [comment]
@praveendvd remove the object from the snapshot

**Comment by praveendvd** [comment]
@christian-bromann  The failure in mac is not this pr related i guess

**Comment by praveendvd** [comment]
@christian-bromann  this error is only for this pull request i am not able to understand what the error means. It just says failed

File: packages/wdio-cli/src/utils.ts [hunk 16]
```diff
@@ -222,7 +222,13 @@ export async function renderConfigurationFile (answers: ParsedAnswers) {
...
     const filename = `wdio.conf.${answers.isUsingTypeScript ? 'ts' : 'js'}`
     const renderedTpl = await renderFile(tplPath, { answers })
-    return fs.promises.writeFile(path.join(process.cwd(), filename), renderedTpl)
+    return fs.promises.writeFile(
+        path.join(
+            process.cwd(),
+            answers.isUsingTypeScript ? 'test' : '', filename
+        ),
+        renderedTpl
+    )
```
Comment (christian-bromann) [inline packages/wdio-cli/src/utils.ts:231]: I think we want to revert this change because we don't want the wdio.conf being generated within `/test/wdio.conf.ts`.

Comment (praveendvd) [reply packages/wdio-cli/src/utils.ts:231]: Hi @christian that was the discussion in previous comment to keep both config and tsconfig in test folder

Comment (christian-bromann) [reply packages/wdio-cli/src/utils.ts:231]: This must be a misunderstanding then. I remember we decided to have the `tsconfig.json` within the test folder but keep `wdio.conf.ts` in the root directory. With this change the current wdio.conf having this config:

```ts
        tsNodeOpts: {
            transpileOnly: true,
            project: 'test/tsconfig.json' // <-- wdio.conf needs to be in the root directory.
        }
```

wouldn't work. Therefore let's keep the wdio.conf in the root directory and the tsconfig.json in /test

Comment (praveendvd) [reply packages/wdio-cli/src/utils.ts:231]: Hi @chritian , the IDE will throw error if tsconfig and wdio config is not in the same directory . That's why we decided keep both in test folder for typescript

Comment (praveendvd) [reply packages/wdio-cli/src/utils.ts:231]: @christian-bromann  What do you think ? :(

**Review by christian-bromann — APPROVED** [review]
Thanks a lot 👍
